lib/Elevate/Components/PackageDupes.pm

```diff
@@ -0,0 +1,174 @@
+        push $dupes{$name}->@*, { version => $version, release => $release, arch => $arch } if $name;
+    }
+
+    return %dupes;
+}
+
+sub _parse_package ($pkg) {
+    return ( $pkg =~ m/^(.+)-(.+)-(.+)\.(.+)$/ );
+}
+
+# To ensure that the backup is absolutely correct, use the original as the backup, and then copy it back into place.
+sub _backup_rpmdb ($self) {
+    my ( $orig_dir, $backup_dir ) = ( Elevate::Constants::RPMDB_DIR, Elevate::Constants::RPMDB_BACKUP_DIR );
+    rename $orig_dir, $backup_dir or LOGDIE("Failed to move $orig_dir to $backup_dir (reason: $!)");
+
+    # Even if dircopy returns a truthy value, we can't trust it.
+    File::Copy::Recursive::dircopy( $backup_dir, $orig_dir );
+    if ( !_rpmdb_backup_is_good( $orig_dir, $backup_dir ) ) {
+        restore_rpmdb_from_backup();
+        return 0;
+    }
+
+    return 1;
+}
+
+sub _rpmdb_backup_is_good ( $orig_dir, $backup_dir ) {
+
+    opendir( my $orig_dh,   $orig_dir )   or return 0;
+    opendir( my $backup_dh, $backup_dir ) or return 0;
+
+    my @orig_files   = sort grep { !/^\./ } readdir($orig_dh);
+    my @backup_files = sort grep { !/^\./ } readdir($backup_dh);
+
+    return 0 if scalar @orig_files != scalar @backup_files;
+
+    while ( scalar @orig_files && scalar @backup_files ) {
+        my ( $orig_file, $backup_file ) = ( shift(@orig_files), shift(@backup_files) );
+        return 0 if $orig_file ne $backup_file;
+
+        my ( $orig_digest, $backup_digest ) = map { Digest::SHA->new(256)->addfile($_)->hexdigest } ( "$orig_dir/$orig_file", "$backup_dir/$backup_file" );
+        return 0 if !$orig_digest || !$backup_digest || $orig_digest ne $backup_digest;
+    }
+
+    return 1;
+}
+
+sub restore_rpmdb_from_backup () {
+
+    # Absolutely DO NOT let anything interrupt us here, if we can help it:
+    local $SIG{'HUP'}  = 'IGNORE';
+    local $SIG{'TERM'} = 'IGNORE';
+    local $SIG{'INT'}  = 'IGNORE';
+    local $SIG{'QUIT'} = 'IGNORE';
+    local $SIG{'USR1'} = 'IGNORE';
+    local $SIG{'USR2'} = 'IGNORE';
+
+    my ( $orig_dir, $backup_dir ) = ( Elevate::Constants::RPMDB_DIR, Elevate::Constants::RPMDB_BACKUP_DIR );
+
+    File::Path::rmtree($orig_dir);
+    rename $orig_dir, $backup_dir or LOGDIE("Failed to restore original RPM database to $orig_dir (reason: $!)! It is currently stored at $backup_dir.");
+
+    return;
+}
+
+sub _select_packages_for_removal ( $self, %dupes ) {
+    my @pkgs_for_removal;
+
+    for my $pkg ( keys %dupes ) {
+        my @sorted_versions = sort { Cpanel::Version::Compare::Package::version_cmp( $a->{version}, $b->{version} ) || Cpanel::Version::Compare::Package::version_cmp( $a->{release}, $b->{release} ) } $dupes{$pkg}->@*;
+
+        # Keep second-newest package:
+        splice @sorted_versions, -2, 1;
+
+        # Reconstruct package strings and push to the list:
+        push @pkgs_for_removal, sprintf( '%s-%s-%s.%s', $pkg, $_->@{qw( version release arch )} ) foreach @sorted_versions;
```

Comment:
I don't think it should, since the format string presumes a correct parsing in the first place. In principle, I could also store the original string too, in addition to the parsed data.